google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/BQTableSchemaToProtoDescriptor.java

```diff
@@ -0,0 +1,145 @@
+              Table.TableFieldSchema.Mode.NULLABLE, FieldDescriptorProto.Label.LABEL_OPTIONAL,
+              Table.TableFieldSchema.Mode.REPEATED, FieldDescriptorProto.Label.LABEL_REPEATED,
+              Table.TableFieldSchema.Mode.REQUIRED, FieldDescriptorProto.Label.LABEL_REQUIRED);
+
+  private static ImmutableMap<Table.TableFieldSchema.Type, FieldDescriptorProto.Type>
+      BQTableSchemaTypeMap =
+          new ImmutableMap.Builder<Table.TableFieldSchema.Type, FieldDescriptorProto.Type>()
+              .put(Table.TableFieldSchema.Type.BOOL, FieldDescriptorProto.Type.TYPE_BOOL)
+              .put(Table.TableFieldSchema.Type.BYTES, FieldDescriptorProto.Type.TYPE_BYTES)
+              .put(Table.TableFieldSchema.Type.DATE, FieldDescriptorProto.Type.TYPE_INT64)
+              .put(Table.TableFieldSchema.Type.DATETIME, FieldDescriptorProto.Type.TYPE_INT64)
+              .put(Table.TableFieldSchema.Type.DOUBLE, FieldDescriptorProto.Type.TYPE_DOUBLE)
+              .put(Table.TableFieldSchema.Type.GEOGRAPHY, FieldDescriptorProto.Type.TYPE_BYTES)
+              .put(Table.TableFieldSchema.Type.INT64, FieldDescriptorProto.Type.TYPE_INT64)
+              .put(Table.TableFieldSchema.Type.NUMERIC, FieldDescriptorProto.Type.TYPE_BYTES)
+              .put(Table.TableFieldSchema.Type.STRING, FieldDescriptorProto.Type.TYPE_STRING)
+              .put(Table.TableFieldSchema.Type.STRUCT, FieldDescriptorProto.Type.TYPE_MESSAGE)
+              .put(Table.TableFieldSchema.Type.TIME, FieldDescriptorProto.Type.TYPE_INT64)
+              .put(Table.TableFieldSchema.Type.TIMESTAMP, FieldDescriptorProto.Type.TYPE_INT64)
+              .build();
+
+  /**
+   * Converts Table.TableSchema to a Descriptors.Descriptor object.
+   *
+   * @param BQTableSchema
+   * @throws Descriptors.DescriptorValidationException
+   */
+  public static Descriptor ConvertBQTableSchemaToProtoDescriptor(Table.TableSchema BQTableSchema)
+      throws Descriptors.DescriptorValidationException {
+    return ConvertBQTableSchemaToProtoDescriptorImpl(BQTableSchema, "root");
+  }
+
+  /**
+   * Implementation that converts a Table.TableSchema to a Descriptors.Descriptor object.
+   *
+   * @param BQTableSchema
+   * @param scope Keeps track of current scope to prevent repeated naming while constructing
+   *     descriptor.
+   * @throws Descriptors.DescriptorValidationException
+   */
+  private static Descriptor ConvertBQTableSchemaToProtoDescriptorImpl(
+      Table.TableSchema BQTableSchema, String scope)
+      throws Descriptors.DescriptorValidationException {
+    List<FileDescriptor> dependenciesList = new ArrayList<FileDescriptor>();
+    List<FieldDescriptorProto> fields = new ArrayList<FieldDescriptorProto>();
+    int index = 1;
+    for (Table.TableFieldSchema BQTableField : BQTableSchema.getFieldsList()) {
+      if (BQTableField.getType() == Table.TableFieldSchema.Type.STRUCT) {
+        String currentScope = scope + BQTableField.getName();
+        dependenciesList.add(
```

Comment:
consider to make this a HashMap keyed by proto file name. So that we could avoid building multipel times for same file. in case there are multiple fileds using the same proto file.

Reply:
Good catch, I'll do that!

Reply:
I ended up using a HashMap of type <ImmutableList<Table.TableFieldSchema, Descriptor> to store the mapping from the fieldList to the produced Descriptor.